compiler/rustc_const_eval/src/errors.rs

```diff
@@ -476,236 +470,509 @@ fn bad_pointer_message(msg: CheckInAllocMsg, handler: &Handler) -> String {
-            RemainderByZero => const_eval_remainder_by_zero,
-            DivisionOverflow => const_eval_division_overflow,
-            RemainderOverflow => const_eval_remainder_overflow,
-            PointerArithOverflow => const_eval_pointer_arithmetic_overflow,
-            InvalidMeta(InvalidMetaKind::SliceTooBig) => const_eval_invalid_meta_slice,
-            InvalidMeta(InvalidMetaKind::TooBig) => const_eval_invalid_meta,
-            UnterminatedCString(_) => const_eval_unterminated_c_string,
-            PointerUseAfterFree(_, _) => const_eval_pointer_use_after_free,
-            PointerOutOfBounds { ptr_size: Size::ZERO, .. } => const_eval_zst_pointer_out_of_bounds,
-            PointerOutOfBounds { .. } => const_eval_pointer_out_of_bounds,
-            DanglingIntPointer(0, _) => const_eval_dangling_null_pointer,
-            DanglingIntPointer(_, _) => const_eval_dangling_int_pointer,
-            AlignmentCheckFailed { .. } => const_eval_alignment_check_failed,
-            WriteToReadOnly(_) => const_eval_write_to_read_only,
-            DerefFunctionPointer(_) => const_eval_deref_function_pointer,
-            DerefVTablePointer(_) => const_eval_deref_vtable_pointer,
-            InvalidBool(_) => const_eval_invalid_bool,
-            InvalidChar(_) => const_eval_invalid_char,
-            InvalidTag(_) => const_eval_invalid_tag,
-            InvalidFunctionPointer(_) => const_eval_invalid_function_pointer,
-            InvalidVTablePointer(_) => const_eval_invalid_vtable_pointer,
-            InvalidStr(_) => const_eval_invalid_str,
-            InvalidUninitBytes(None) => const_eval_invalid_uninit_bytes_unknown,
-            InvalidUninitBytes(Some(_)) => const_eval_invalid_uninit_bytes,
-            DeadLocal => const_eval_dead_local,
-            ScalarSizeMismatch(_) => const_eval_scalar_size_mismatch,
-            UninhabitedEnumVariantWritten(_) => const_eval_uninhabited_enum_variant_written,
-            UninhabitedEnumVariantRead(_) => const_eval_uninhabited_enum_variant_read,
-            ValidationError(e) => e.diagnostic_message(),
-            Custom(x) => (x.msg)(),
+
+        let handler = &self.tcx.sess.parse_sess.span_diagnostic;
+
+        match self.err {
+            #[allow(rustc::untranslatable_diagnostic)]
+            Ub(str) => {
+                eagerly_translate_with_args!(handler, str.clone().into(),)
+            }
+            Unreachable => {
+                eagerly_translate_with_args!(handler, const_eval_unreachable,)
+            }
+            BoundsCheckFailed { len, index } => {
+                eagerly_translate_with_args!(
+                    handler,
+                    const_eval_bounds_check_failed,
+                    "len",
+                    len,
+                    "index",
+                    index,
+                )
+            }
+            DivisionByZero => {
+                eagerly_translate_with_args!(handler, const_eval_division_by_zero,)
+            }
```

Comment:
There's a lot of code duplication and I think we should avoid that. Perhaps this should be an associated function in `UndefinedBehaviorInfoExt` that uses a hack similar to what I have done in miri:

https://github.com/rust-lang/rust/blob/b08dd92552d663e3c877c8e5ce859e212205a09f/src/tools/miri/src/diagnostics.rs#L339-L347

Reply:
Are you saying this because of the `*Ext` and `*Ext2` variants both existing? `*Ext` is the old implementation which I will remove.

Reply:
I.. don't think that will work out? But feel free to try. I'm mostly thinking that what `*Ext2` is renamed to should mostly use `*Ext`'s code.

Reply:
`*Ext` types are no longer needed, they are an `IntoDiagnostic` implementation and can be completely removed since the current implementation doesn't use `IntoDiagnostic`.

Reply:
Ah okay I see. Go ahead and remove it.